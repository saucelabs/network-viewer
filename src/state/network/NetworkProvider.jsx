import React, { useEffect, useReducer, useMemo } from 'react';
import PropTypes from 'prop-types';

import { reducer, initialState as defaultState } from './reducer';
import { updateData, fetchFile, updateScrollToIndex } from './actions';
import { NetworkContext } from './Context';
import { findRequestIndex } from '../../utils';
import { ROW_ID_PREFIX } from '../../constants';

const NetworkProvider = (props) => {
  const {
    data,
    file,
    fetchOptions,
    initialState,
    scrollTimeStamp,
    scrollRequestPosition,
    autoHighlightChange,
    onDataLoaded,
    onDataError,
    onPause,
    onResume,
    onReset,
  } = props;

  const [state, dispatch] = useReducer(reducer, initialState);
  const callbacks = {
    onPause,
    onResume,
    onReset,
  };
  const value = useMemo(() => [state, dispatch, callbacks], [state]);
  const selectedReqIndex = value[0].get('selectedReqIndex');
  const requestData = value[0].get('data');
  const showReqDetail = value[0].get('showReqDetail');
  const actualData = value[0].get('actualData');
  const error = value[0].get('error');

  // Update data onChange of network data
  useEffect(() => {
    if (data && data.log && data.log.entries) {
      updateData(dispatch)(data);
    }
  }, [data]);

  // Fetch HAR file onChange of file prop
  useEffect(() => {
    if (file) {
      fetchFile(dispatch)(file, fetchOptions);
    }
  }, [file]);

  useEffect(() => {
    if (actualData.size && onDataLoaded) {
      onDataLoaded(actualData);
    }
  }, [actualData]);

  useEffect(() => {
    if (error && onDataError) {
      onDataError(error);
    }
  }, [error]);

  // Find nearby request-rowId and update scrollIndex on scrollTimeStamp receive
  useEffect(() => {
    const shouldChangeHighlight = showReqDetail ? autoHighlightChange : true;
    if (scrollTimeStamp && shouldChangeHighlight) {
      const reqIndex = findRequestIndex({
        data: requestData,
        timestamp: scrollTimeStamp,
        position: scrollRequestPosition,
      });
      if (reqIndex || reqIndex === 0) {
        updateScrollToIndex(dispatch)(requestData.get(reqIndex));
      }
    }
  }, [scrollTimeStamp, actualData]);

  // Scroll to request row onChange of scrollToIndex
  // setTimeout is required when reqDetail is visible,
  // it allows DOM to adjust before we scroll it to highlighted request
  useEffect(() => {
    if (selectedReqIndex) {
      setTimeout(() => {
        const element = document.getElementById(ROW_ID_PREFIX + selectedReqIndex);
        if (element) {
          element.scrollIntoView({
            alignToTop: true,
            behavior: 'smooth',
          });
        }
      }, 300);
    }
  }, [selectedReqIndex]);

  return (
    <NetworkContext.Provider
      value={value}
      {...props}
    />
  );
};

NetworkProvider.propTypes = {
  autoHighlightChange: PropTypes.bool,
  data: PropTypes.object,
  fetchOptions: PropTypes.object,
  file: PropTypes.string,
  initialState: PropTypes.object,
  onDataError: PropTypes.func,
  onDataLoaded: PropTypes.func,
  onPause: PropTypes.func,
  onReset: PropTypes.func,
  onResume: PropTypes.func,
  scrollRequestPosition: PropTypes.oneOf(['before', 'after', 'near']),
  scrollTimeStamp: PropTypes.number,
};

NetworkProvider.defaultProps = {
  autoHighlightChange: false,
  data: null,
  fetchOptions: { withCredentials: true },
  file: null,
  initialState: defaultState,
  onDataError: null,
  onDataLoaded: null,
  onPause: null,
  onReset: null,
  onResume: null,
  scrollRequestPosition: 'near',
  scrollTimeStamp: null,
};

export default NetworkProvider;
