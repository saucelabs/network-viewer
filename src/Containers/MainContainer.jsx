import React from 'react';
import PropTypes from 'prop-types';

import { useNetwork } from '../state/network/Context';
import FilterContainer from './FilterContainer';
import NetworkTableContainer from './NetworkTableContainer';
import LoaderContainer from './../Components/LoaderContainer';
import { FETCH_FILE_LOAD_TEXT } from '../constants';
import ReqDetailContainer from './ReqDetailContainer';
import Styles from './MainContainer.styles.scss';
import TimelineContainer from './TimelineContainer';
import { useTheme } from '../state/theme/Context';
import NetworkTableFooter from './../Components/NetworkTable/NetworkTableFooter';

const MainContainer = ({ onRequestSelect }) => {
  const { state } = useNetwork();
  const { showTimeline } = useTheme();
  const loading = state.get('loading');
  const showReqDetail = state.get('showReqDetail');
  const dataSummary = state.get('dataSummary');
  const actualData = state.get('actualData');

  return (
    <>
      <LoaderContainer
        show={loading}
        text={FETCH_FILE_LOAD_TEXT}
      >
        {showTimeline && <TimelineContainer />}
        <FilterContainer />
        <section className={Styles['main-container']}>
          <NetworkTableContainer onRequestSelect={onRequestSelect} />
          {showReqDetail && <ReqDetailContainer />}
        </section>
        {actualData.size ? <NetworkTableFooter dataSummary={dataSummary} /> : null}
      </LoaderContainer>
    </>
  );
};

MainContainer.propTypes = {
  onRequestSelect: PropTypes.func,
};

MainContainer.defaultProps = {
  onRequestSelect: () => {},
};

export default MainContainer;
