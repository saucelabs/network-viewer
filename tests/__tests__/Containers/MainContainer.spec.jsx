import React from 'react';
import { mount } from 'enzyme';

import MainContainer from './../../../src/Containers/MainContainer';
import NetworkProvider from './../../../src/state/network/NetworkProvider';

describe('MainContainer', () => {
  it('renders without crashing', () => {
    const element = mount(
      <NetworkProvider>
        <MainContainer />
      </NetworkProvider>,
    );
    expect(element).toMatchSnapshot();
  });
});
