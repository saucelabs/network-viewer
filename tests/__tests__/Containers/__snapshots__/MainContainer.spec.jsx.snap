// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainContainer renders without crashing 1`] = `
<NetworkProvider
  autoHighlightChange={false}
  data={null}
  fetchOptions={
    Object {
      "withCredentials": true,
    }
  }
  file={null}
  initialState={
    Immutable.Map {
      "errorFilter": false,
      "search": Object {
        "name": "url",
        "value": "",
      },
      "error": null,
      "scrollToIndex": null,
      "data": Immutable.List [],
      "selectedReqIndex": null,
      "reqDetail": null,
      "showReqDetail": false,
      "totalNetworkTime": null,
      "actualData": Immutable.List [],
      "filter": Object {
        "name": null,
        "value": null,
      },
      "dataSummary": Immutable.Map {},
      "loading": false,
      "sort": Object {
        "isAcs": true,
        "key": "startedDateTime",
      },
    }
  }
  onDataError={null}
  onDataLoaded={null}
  scrollRequestPosition="near"
  scrollTimeStamp={null}
>
  <MainContainer
    onRequestSelect={[Function]}
  >
    <LoaderContainer
      show={false}
      text="Please wait, Fetching file is in progress."
    >
      <FilterContainer>
        <section>
          <div>
            <Search
              name="url"
              value=""
            >
              <input
                onChange={[Function]}
                placeholder="Search by full URL"
                type="text"
                value=""
              />
            </Search>
          </div>
          <div>
            <Button
              className="filter-button selected-filter"
              disabled={false}
              href={null}
              key="All"
              onClick={[Function]}
              size="sm"
              type="button"
              variant="default"
            >
              <button
                className="btn filter-button selected-filter"
                disabled={false}
                href={null}
                onClick={[Function]}
                size="sm"
                type="button"
              >
                All
              </button>
            </Button>
            <Button
              className="filter-button"
              disabled={false}
              href={null}
              key="XHR"
              onClick={[Function]}
              size="sm"
              type="button"
              variant="default"
            >
              <button
                className="btn filter-button"
                disabled={false}
                href={null}
                onClick={[Function]}
                size="sm"
                type="button"
              >
                XHR
              </button>
            </Button>
            <Button
              className="filter-button"
              disabled={false}
              href={null}
              key="JS"
              onClick={[Function]}
              size="sm"
              type="button"
              variant="default"
            >
              <button
                className="btn filter-button"
                disabled={false}
                href={null}
                onClick={[Function]}
                size="sm"
                type="button"
              >
                JS
              </button>
            </Button>
            <Button
              className="filter-button"
              disabled={false}
              href={null}
              key="CSS"
              onClick={[Function]}
              size="sm"
              type="button"
              variant="default"
            >
              <button
                className="btn filter-button"
                disabled={false}
                href={null}
                onClick={[Function]}
                size="sm"
                type="button"
              >
                CSS
              </button>
            </Button>
            <Button
              className="filter-button"
              disabled={false}
              href={null}
              key="Img"
              onClick={[Function]}
              size="sm"
              type="button"
              variant="default"
            >
              <button
                className="btn filter-button"
                disabled={false}
                href={null}
                onClick={[Function]}
                size="sm"
                type="button"
              >
                Img
              </button>
            </Button>
            <Button
              className="filter-button"
              disabled={false}
              href={null}
              key="Media"
              onClick={[Function]}
              size="sm"
              type="button"
              variant="default"
            >
              <button
                className="btn filter-button"
                disabled={false}
                href={null}
                onClick={[Function]}
                size="sm"
                type="button"
              >
                Media
              </button>
            </Button>
            <Button
              className="filter-button"
              disabled={false}
              href={null}
              key="Font"
              onClick={[Function]}
              size="sm"
              type="button"
              variant="default"
            >
              <button
                className="btn filter-button"
                disabled={false}
                href={null}
                onClick={[Function]}
                size="sm"
                type="button"
              >
                Font
              </button>
            </Button>
            <Button
              className="filter-button"
              disabled={false}
              href={null}
              key="Doc"
              onClick={[Function]}
              size="sm"
              type="button"
              variant="default"
            >
              <button
                className="btn filter-button"
                disabled={false}
                href={null}
                onClick={[Function]}
                size="sm"
                type="button"
              >
                Doc
              </button>
            </Button>
            <Button
              className="filter-button"
              disabled={false}
              href={null}
              key="WS"
              onClick={[Function]}
              size="sm"
              type="button"
              variant="default"
            >
              <button
                className="btn filter-button"
                disabled={false}
                href={null}
                onClick={[Function]}
                size="sm"
                type="button"
              >
                WS
              </button>
            </Button>
            <Button
              className="filter-button"
              disabled={false}
              href={null}
              key="Manifest"
              onClick={[Function]}
              size="sm"
              type="button"
              variant="default"
            >
              <button
                className="btn filter-button"
                disabled={false}
                href={null}
                onClick={[Function]}
                size="sm"
                type="button"
              >
                Manifest
              </button>
            </Button>
            <ErrorFilter
              isError={false}
              onChange={[Function]}
            >
              <Checkbox
                containerClassName=""
                isChecked={false}
                onChange={[Function]}
                title="error"
              >
                <div
                  className="checkbox-container"
                >
                  <label
                    title="error"
                  >
                    <input
                      checked={false}
                      onChange={[Function]}
                      type="checkbox"
                    />
                    Errors Only
                  </label>
                </div>
              </Checkbox>
            </ErrorFilter>
            <ImportHar
              className={null}
              showButton={true}
            >
              <div
                onBlur={[Function]}
                onClick={[Function]}
                onDragEnter={[Function]}
                onDragLeave={[Function]}
                onDragOver={[Function]}
                onDrop={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                tabIndex={0}
              >
                <input
                  accept=".har"
                  autoComplete="off"
                  multiple={false}
                  onChange={[Function]}
                  onClick={[Function]}
                  style={
                    Object {
                      "display": "none",
                    }
                  }
                  tabIndex={-1}
                  type="file"
                />
                <Button
                  className={null}
                  disabled={false}
                  href={null}
                  type="button"
                  variant="default"
                >
                  <button
                    className="btn"
                    disabled={false}
                    href={null}
                    type="button"
                  >
                    Import HAR
                  </button>
                </Button>
              </div>
            </ImportHar>
            <Reset
              className=""
              onReset={[Function]}
            >
              <Button
                className=""
                disabled={false}
                href={null}
                onClick={[Function]}
                type="button"
                variant="default"
              >
                <button
                  className="btn"
                  disabled={false}
                  href={null}
                  onClick={[Function]}
                  type="button"
                >
                  Reset
                </button>
              </Button>
            </Reset>
          </div>
        </section>
      </FilterContainer>
      <section>
        <NetworkTableContainer
          onRequestSelect={[Function]}
        >
          <section>
            <ImportHar
              className={null}
              showButton={false}
            >
              <div
                onBlur={[Function]}
                onClick={[Function]}
                onDragEnter={[Function]}
                onDragLeave={[Function]}
                onDragOver={[Function]}
                onDrop={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                tabIndex={0}
              >
                <input
                  accept=".har"
                  autoComplete="off"
                  multiple={false}
                  onChange={[Function]}
                  onClick={[Function]}
                  style={
                    Object {
                      "display": "none",
                    }
                  }
                  tabIndex={-1}
                  type="file"
                />
                <p>
                  Drag and drop HAR file here, or click to select file
                </p>
              </div>
            </ImportHar>
            <InputHAR>
              <div>
                <h4>
                  OR add HAR file URL in the below input box
                </h4>
                <URLInput
                  onSubmit={[Function]}
                >
                  <div>
                    <CORSCheckbox
                      isEnabled={false}
                      onChange={[Function]}
                    >
                      <Checkbox
                        containerClassName=""
                        isChecked={false}
                        onChange={[Function]}
                        title="CORS Enable"
                      >
                        <div
                          className="checkbox-container"
                        >
                          <label
                            title="CORS Enable"
                          >
                            <input
                              checked={false}
                              onChange={[Function]}
                              type="checkbox"
                            />
                            CORS
                          </label>
                        </div>
                      </Checkbox>
                    </CORSCheckbox>
                    <input
                      onChange={[Function]}
                      placeholder="HAR file URL"
                      type="text"
                      value=""
                    />
                    <Button
                      className={null}
                      disabled={false}
                      href={null}
                      onClick={[Function]}
                      type="button"
                      variant="default"
                    >
                      <button
                        className="btn"
                        disabled={false}
                        href={null}
                        onClick={[Function]}
                        type="button"
                      >
                        GO
                      </button>
                    </Button>
                  </div>
                </URLInput>
                <p>
                  <span>
                    For Example use this har file
                  </span>
                  <a
                    href="https://raw.githubusercontent.com/saucelabs/network-viewer/main/examples/src/data/network.har"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    https://raw.githubusercontent.com/saucelabs/network-viewer/main/examples/src/data/network.har
                  </a>
                </p>
              </div>
            </InputHAR>
          </section>
        </NetworkTableContainer>
      </section>
    </LoaderContainer>
  </MainContainer>
</NetworkProvider>
`;
